feat: markdown config typechecking

JuanM04 opened the pull request:
## Changes

Following the comment `// TODO: add better type checking`, I've added type checking for the markdown configuration

## Testing

<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

## Docs

<!-- Is this a visible change? You probably need to update docs! -->
<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

FredKSchott commented on packages/astro/src/@types/astro.ts:
```diff
@@ -349,11 +349,11 @@ export interface AstroUserConfig {
 		 * @docs
 		 * @name markdown.shikiConfig
-		 * @type {ShikiConfig}
+		 * @type {Partial<ShikiConfig>}
```
can you change to `@typeraw`? I think this would error as invalid JSDoc

FredKSchott commented on packages/astro/src/core/config.ts:
```diff
@@ -12,6 +13,9 @@ import loadTypeScript from '@proload/plugin-tsm';
 import postcssrc from 'postcss-load-config';
 import { arraify, isObject } from './util.js';
 import { appendForwardSlash, trimSlashes } from './path.js';
+// These two imports make astroMarkdownOptions work
```
wahhh??? Would love an explanation why here, for future readers

JuanM04 replied:
Check out the new commit, I added a longer (not sure if better) explanation

FredKSchott commented on packages/astro/src/core/config.ts:
```diff
@@ -138,25 +142,7 @@ export const AstroConfigSchema = z.object({
-		.passthrough()
-		.optional()
-		.default({}),
+	markdown: astroMarkdownOptions.default({}).default({}),
```
only need 1 `default()` call

natemoo-re commented on packages/astro/src/@types/astro.ts:
```diff
@@ -349,14 +349,24 @@ export interface AstroUserConfig {
 		 */
 		drafts?: boolean;
 
+		/**
+		 * @docs
+		 * @name markdown.mode
+		 * @type {'md' | 'mdx'}
+		 * @default `mdx`
+		 * @description
+		 * Control wheater to allow components inside markdown files ('mdx') or not ('md').
+		 */
+		mode?: 'md' | 'mdx';
+
```
I don't think this actually does anything right now, which is why we didn't include it!

JuanM04 replied:
It does add `rehypeJsx` and `rehypeExpressions`. I could remove the MDX mode altogether if these plugins do nothing

natemoo-re replied:
Ah no you're right! I also saw that https://github.com/withastro/astro/pull/2971 added escaping for expressions when `mode === 'md'`. So I think this is fine!

JuanM04 replied:
After doing some testing, the only difference I've found is that components are being rendered with `mdx` but completely ignored with `md`. Expressions works in both modes, tho.

natemoo-re reviewed (changes requested):
I'd remove any public references to the `mode` option for now, since it doesn't do anything.

Other than that, looks great!

FredKSchott reviewed (changes requested):
Marking as red just to make sure we hold off on this until post release day. I have some comments that I'd like to make but just won't have the time until Monday/Tuesday to make them. Appreciate the patience!

FredKSchott commented on packages/astro/src/core/config.ts:
```diff
@@ -140,27 +147,7 @@ export const AstroConfigSchema = z.object({
-		.passthrough()
-		.optional()
-		.default({}),
+	markdown: astroMarkdownOptions.default({}),
```
Thanks again for your patience on this review @JuanM04! 

My main comment is that we really benefit from having all of our config parsing logic here in one file and in one single parser object instantiation. 

Reading this PR, I can see the value you've put into validation reuse, which is great. But, I'm also bouncing around to follow where the different parsers / validation logic lives, defined across multiple files, and even multiple packages, with multiple zod parser instances referencing each other.

If you'd like to improve the existing `markdown: z.object(...` validation logic defined here, I'm all for it. I could even see the value in a new `Plugin` validator that both `remarkPlugins` and `rehypePlugins`, which breaks up the single-object-validator that I love but for a valid, valuable reason.

But at the end of the day, Astro owns the "Astro Config" type and therefore its validation, so I'd like to keep it as simple as possible and only inside the Astro package, for our sake as maintainers.

@JuanM04 lmk what you think, and how you'd like to move forward!

JuanM04 replied:
You're totally right. The idea behind the division was about separation of concerns, since `@astrojs/markdown-remark` is its own package, but I see how that can be a problem.

I'll integrate this configuration in a single place

JuanM04 commented:
@FredKSchott I've moved the validation inside `config.ts`. Also, I've removed all the internal references to the old `markdownRender` way of importing markdown, so it's cleaner and doesn't do unnecessary work

JuanM04 commented:
Note: the error thrown by Deno it's caused by this PR directly. If you have an Astro page with the `<Markdown />` component and build it with Deno, it will throw "No Markdown parser found"

FredKSchott reviewed (approved):
Nice! This is looking great, thanks for making those changes. LGTM!

FredKSchott commented:
> Note: the error thrown by Deno it's not caused by this PR directly. If you have an Astro page with the `<Markdown />` component and build it with Deno, it will throw "No Markdown parser found"

Can't merge with failing tests! What's the best way to resolve?

JuanM04 commented:
> Can't merge with failing tests! What's the best way to resolve?

I'm looking into it. Either a fix is found or the test is ignored

FredKSchott commented:
As long as you can confirm that this isn't a regression caused by this PR

JuanM04 commented:
Ok, this way it should work. The plugin now is dynamically imported only when using `<Markdown />` _and_ Node.js. When using Deno, it throws an error

FredKSchott reviewed (approved):
LGTM!
